File: ccc/2020/s3.py
ALPHABET = "abcdefghijklmnopqrstuvwxyz"

needle = input()
haystack = input()

needle_size = len(needle)
haystack_size = len(haystack)
discovered_needles = set([])

p = 31
m = 10**9 + 9


def hash_string(string):
    p_pow = 1
    hash_value = 0

    for char in string:
        hash_value = (hash_value + ord(char) * p_pow) % m
        p_pow = (p_pow * p) % m

    return hash_value


needle_count = [0] * 26
for letter in needle:
    needle_count[ord(letter) - 97] += 1

first_needle = {}

letter_count = [0] * 26

for i in range(haystack_size - (needle_size - 1)):
    possible_needle = haystack[i : i + needle_size]

    if i == 0:
        for letter in possible_needle:
            letter_count[ord(letter) - 97] += 1
    else:
        letter_count[ord(haystack[i - 1]) - 97] -= 1
        letter_count[ord(haystack[i + needle_size - 1]) - 97] += 1

    if letter_count != needle_count:
        continue

    hash_value = hash(possible_needle)

    discovered_needles.add(hash_value)

print(len(discovered_needles))

File: ccc/2020/s3.cpp
#include <set>
#include <array>
#include <string>
#include <iostream>

using namespace std;

const int p = 31;
const int m = 1e9 + 9;
long long hasher(string s) {
    long long hash_value = 0;
    long long p_pow = 1;
    for (char c : s) {
        hash_value = (hash_value + (c - 'a' + 1) * p_pow) % m;
        p_pow = (p_pow * p) % m;
    }

    return hash_value;
}

int main() {
    string needle, haystack;
    cin >> needle >> haystack;

    set<long long> matches;
    array<int, 26> window_freq = {};
    array<int, 26> needle_freq = { 0 };

    for (char c : needle) {
        needle_freq[c - 'a']++;
    }

    for (int i = 0; i < haystack.size() - needle.size() + 1; i++) {
        string possible_needle = haystack.substr(i, needle.size());

        if (i == 0) {
            for (char c : possible_needle) {
                window_freq[c - 'a']++;
            }
        } else {
            window_freq[haystack[i - 1] - 'a']--;
            window_freq[haystack[i + needle.size() - 1] - 'a']++;
        }

        if (needle_freq == window_freq) {
            matches.insert(hasher(possible_needle));
        }   
    }

    cout << matches.size() << endl;
}